File: manual/data/python/src/function_nltk_chartparser.py
import nltk

class TestChartParser:

    def test_chartparser(self, sentence):
        grammar = nltk.CFG.fromstring("""
            S -> NP VP
            NP -> Det N
            VP -> V NP
            Det -> 'the'
            N -> 'dog' | 'cat'
            V -> 'chased'
        """)
        parser = nltk.ChartParser(grammar)
        # 返回解析树的字符串形式（所有树合并为一个字符串，避免返回列表）
        parsed_trees = [str(tree) for tree in parser.parse(sentence.split())]
        if parsed_trees:
            return ' | '.join(parsed_trees)  # 用竖线分隔多个解析树
        else:
            return ''

File: manual/data/cpp/src/function_nltk_chartparser.cpp
#include <string>
#include <vector>
#include <sstream>
#include <boost/spirit/include/qi.hpp>
#include <boost/phoenix.hpp>

class TestChartParser {
public:
    std::vector<std::string> test_chartparser(const std::string& sentence) {
        namespace qi = boost::spirit::qi;
        namespace ascii = boost::spirit::ascii;

        std::vector<std::string> results;
        std::string::const_iterator iter = sentence.begin();
        std::string::const_iterator end = sentence.end();

        bool r = qi::phrase_parse(iter, end,
            (
                qi::lit("the") >> (qi::lit("dog") | qi::lit("cat")) >> 
                qi::lit("chased") >> qi::lit("the") >> (qi::lit("dog") | qi::lit("cat"))
            ),
            ascii::space
        );

        if (r && iter == end) {
            results.push_back("Parse successful");
        }

        return results;
    }
};